### PR TITLE
Add support for PuppetDB API v4

### PR DESCRIPTION
**Pull request by eavgerinos**
- Fix finding nodes by using `certname` over `name` (v3, v4)
- Support non-string facts
- Add `:puppetdb_api_version` option instead of parsing the URI to guess the version

### CODE REVIEW

**Comment by timogoebel** [comment]
@eavgerinos: Thanks.

[test] - this kicks our ci pipeline

**Comment by timogoebel** [comment]
[test]

File: app/models/setting/puppetdb.rb [hunk 1]
```diff
@@ -20,14 +22,16 @@ def self.default_settings
     default_ssl_ca_file ||= (SETTINGS[:ssl_ca_file]).to_s
     default_ssl_certificate ||= (SETTINGS[:ssl_certificate]).to_s
     default_ssl_private_key ||= (SETTINGS[:ssl_priv_key]).to_s
+    default_api_version ||= 4
```
Comment (timogoebel) [inline app/models/setting/puppetdb.rb]: This will break existing installations. Can we calculate the default based on the URL (you removed that in app/services/puppetdb.rb)?

File: app/services/puppetdb.rb [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
+    case api_version
+    when 1
+      PuppetdbClient::V1.new(options)
+    when 3
       PuppetdbClient::V3.new(options)
+    when 4
+      PuppetdbClient::V4.new(options)
     else
-      PuppetdbClient::V1.new(options)
+      raise ::Puppetdb::UnsupportedVersion
```
Comment (timogoebel) [inline app/services/puppetdb.rb]: Please use `raise Foreman::Exception.new(N_('Unsupported PuppetDB version.'))` instead. This will generate an [ERF-Code](http://projects.theforeman.org/projects/foreman/wiki/ErrorCodes) for easier troubleshooting.

File: app/services/puppetdb.rb [hunk 3]
```diff
@@ -1,15 +1,26 @@
...
   def self.client
     options = {
       :uri => uri,
       :ssl_ca_file => Setting[:puppetdb_ssl_ca_file],
       :ssl_certificate_file => Setting[:puppetdb_ssl_certificate],
       :ssl_private_key_file => Setting[:puppetdb_ssl_private_key]
     }
-    if uri.path.start_with?('/pdb')
+
+    api_version = Setting[:puppetdb_api_version].to_i
```
Comment (timogoebel) [inline app/services/puppetdb.rb]: Let's extract this to a method?

File: app/services/puppetdb_host.rb [hunk 4]
```diff
@@ -26,7 +26,7 @@ def parse_facts(pdb_facts)
   end
 
   def parse_fact_value(value)
-    result = JSON.parse(value)
+    result = value.is_a?(String) ? JSON.parse(value) : value
```
Comment (timogoebel) [inline app/services/puppetdb_host.rb]: Why is this necessary?

File: app/models/setting/puppetdb.rb [hunk 5]
```diff
@@ -20,14 +22,16 @@ def self.default_settings
...
 
     [
       set('puppetdb_enabled', _("Integration with PuppetDB, enabled will deactivate a host in PuppetDB when it's deleted in Foreman"), default_enabled),
       set('puppetdb_address', _('Foreman will send PuppetDB requests to this address'), default_address),
       set('puppetdb_dashboard_address', _('Foreman will proxy PuppetDB Performance Dashboard requests to this address'), default_dashboard_address),
       set('puppetdb_ssl_ca_file', _('Foreman will send PuppetDB requests with this CA file'), default_ssl_ca_file),
       set('puppetdb_ssl_certificate', _('Foreman will send PuppetDB requests with this certificate file'), default_ssl_certificate),
-      set('puppetdb_ssl_private_key', _('Foreman will send PuppetDB requests with this key file'), default_ssl_private_key)
+      set('puppetdb_ssl_private_key', _('Foreman will send PuppetDB requests with this key file'), default_ssl_private_key),
+      set('puppetdb_api_version', _('Foreman will use this PuppetDB API version'), default_api_version)
```
Comment (timogoebel) [inline app/models/setting/puppetdb.rb]: Can we make this a select field?

Something like:

```ruby
module Puppetdb
  API_VERSIONS = {
    '4' => 'v4: PuppetDB 2.3, 3.0, 3.1, 3.2, 4.0 (PE 3.8, 2015.2, 2015.3)',
    '3' => 'v3: PuppetDB 1.5, 1.6 (PE 3.1, 3.2, 3.3)',
    '2' => 'v2: PuppetDB 1.1, 1.2, 1.3, 1.4'
}
[...]
end


set('puppetdb_api_version',
            _("Foreman will use this PuppetDB API version"),
            'delete', N_('PuppetDB API version'), nil, {:collection => Proc.new {::Puppetdb::API_VERSIONS} })

File: app/services/puppetdb_host.rb (hunk 4 above)
Comment (eavgerinos) [inline app/services/puppetdb_host.rb]: [Puppet supports non-string facts](https://docs.puppet.com/puppet/4.10/lang_facts_and_builtin_vars.html#data-types) (boolean for example), so in the case of `JSON.parse(false)` we would get a `TypeError`,  in order to avoid this, it's safe to JSON parse only strings.

File: app/models/setting/puppetdb.rb (hunk 1 above)
Comment (eavgerinos) [inline app/models/setting/puppetdb.rb]: I was thinking a major version release. If we want to avoid it, I can re-introduce this logic, but I think the settings module would be the suitable place for it.

What do you think?

File: app/models/setting/puppetdb.rb (hunk 5 above)
Comment (eavgerinos) [inline app/models/setting/puppetdb.rb]: Of course, will do!

File: app/services/puppetdb.rb (hunk 3 above)
Comment (eavgerinos) [inline app/services/puppetdb.rb]: Will do.

File: app/services/puppetdb.rb (hunk 2 above)
Comment (eavgerinos) [inline app/services/puppetdb.rb]: I see! Nice catch

File: app/models/setting/puppetdb.rb (hunk 1 above)
Comment (timogoebel) [inline app/models/setting/puppetdb.rb]: I think, adding a db migration might also work for this. Let me know, if you want me to try that out.

**Comment by timogoebel** [comment]
@eavgerinos: Any news on this? Do you need help somewhere?

**Comment by eavgerinos** [comment]
@timogoebel Sorry, I was busy with other issues, I'll implement the requested changes asap.

File: app/models/setting/puppetdb.rb (hunk 1 above)
Comment (eavgerinos) [inline app/models/setting/puppetdb.rb]: What would the migration do in this case?

Comment (timogoebel) [inline app/models/setting/puppetdb.rb]: How about something like [this](https://github.com/timogoebel/puppetdb_foreman/commit/7c86b403fabe6f14389978b4b7c2332dc942aef6)?

**Comment by timogoebel** [comment]
@eavgerinos : Thanks. Could you check out if https://github.com/timogoebel/puppetdb_foreman/commit/7c86b403fabe6f14389978b4b7c2332dc942aef6 is a suitable solution for you? If you like it, just squash it in and we're ready to merge once tests pass.

**Comment by timogoebel** [comment]
[test]

**Comment by timogoebel** [comment]
Closing due to inactivity.
